plutus-chain-index/src/Plutus/ChainIndex/Emulator/Handlers.hs

```diff
@@ -0,0 +1,131 @@
+    ) => ChainIndexQueryEffect
+    ~> Eff effs
+handleQuery = \case
+    DatumFromHash h -> gets (view $ diskState . dataMap . at h)
+    ValidatorFromHash h -> gets (view $ diskState . validatorMap . at h)
+    MintingPolicyFromHash h -> gets (view $ diskState . mintingPolicyMap . at h)
+    TxOutFromRef TxOutRef{txOutRefId, txOutRefIdx} -> gets @ChainIndexEmulatorState (preview $ diskState . txMap . ix txOutRefId . citxOutputs . ix (fromIntegral txOutRefIdx))
+    TxFromTxId i -> getTxFromTxId i
+    UtxoSetMembership r -> do
+        utxoState <- gets (measure . view utxoIndex)
+        case tip utxoState of
+            Nothing -> throwError QueryFailedNoTip
+            Just tp -> pure (tp, isUnspentOutput r utxoState)
+    UtxoSetAtAddress cred -> do
+        state <- get
+        let outRefs = view (diskState . addressMap . at cred) state
+            utxoState = view (utxoIndex . to measure) state
+            page = pageOf def $ Set.filter (\r -> isUnspentOutput r utxoState) (fromMaybe mempty outRefs)
+        case tip utxoState of
+            Nothing -> throwError QueryFailedNoTip
+            Just tp -> pure (tp, page)
+    GetTip ->
+        gets (tip . measure . view utxoIndex) >>= maybe (throwError QueryFailedNoTip) pure
+
+handleControl ::
+    forall effs.
+    ( Member (State ChainIndexEmulatorState) effs
+    , Member (Error ChainIndexError) effs
+    , Member (LogMsg ChainIndexLog) effs
+    )
+    => ChainIndexControlEffect
+    ~> Eff effs
+handleControl = \case
+    AppendBlock tip_ transactions -> do
+        oldState <- get @ChainIndexEmulatorState
+        case UtxoState.insert (UtxoState.fromBlock tip_ transactions) (view utxoIndex oldState) of
+            Left err -> do
+                let reason = InsertionFailed err
+                logError $ Err reason
+                throwError reason
+            Right InsertUtxoSuccess{newIndex, insertPosition} -> do
+                put $ oldState
+                        & set utxoIndex newIndex
+                        & over diskState (mappend $ foldMap DiskState.fromTx transactions)
+                logDebug $ InsertionSuccess tip_ insertPosition
+    Rollback tip_ -> do
+        oldState <- get @ChainIndexEmulatorState
+        case UtxoState.rollback tip_ (view utxoIndex oldState) of
+            Left err -> do
+                let reason = RollbackFailed err
+                logError $ Err reason
+                throwError reason
+            Right RollbackResult{newTip, rolledBackIndex} -> do
+                put $ oldState & set utxoIndex rolledBackIndex
+                logDebug $ RollbackSuccess newTip
+    CollectGarbage -> do
+        -- Rebuild the index using only transactions that still have at
+        -- least one output in the UTXO set
+        utxos <- gets (Set.toList . Set.map txOutRefId . UtxoState.unspentOutputs . UtxoState.utxoState . view utxoIndex)
+        newDiskState <- foldMap DiskState.fromTx . catMaybes <$> mapM getTxFromTxId utxos
```

Comment:
Should it be an error if `getTxFromTxId` fails here?

Reply:
I changed `getTxFromTxId` to log a warning if the TX wasn't found